server/mqtt.go

```diff
@@ -2503,106 +2710,217 @@ func (sess *mqttSession) update(filters []*mqttFilter, add bool) error {
+		_, usedForPubRel := sess.pendingPubRel[pi]
+		if /*dup && */ usedForPubRel {
+			return 0, false
+		}
+
+		// We should have a pending JS ACK for this PI.
+		ack = sess.pendingPublish[pi]
+	} else {
 		// sess.maxp will always have a value > 0.
-		if len(sess.pending) >= int(sess.maxp) {
+		if len(sess.pendingPublish) >= int(sess.maxp) {
 			// Indicate that we did not assign a packet identifier.
 			// The caller will not send the message to the subscription
 			// and JS will redeliver later, based on consumer's AckWait.
 			return 0, false
 		}
-		pi = bumpPI()
+
+		pi = sess.bumpPI()
+		if pi == 0 {
+			return 0, false
+		}
+
 		sseqToPi[sseq] = pi
 	}
-	if pending == nil {
-		pending = &mqttPending{jsDur: jsDur, sseq: sseq, subject: reply}
-		sess.pending[pi] = pending
-	}
-	// If redelivery, return DUP flag
-	if dcount > 1 {
-		dup = true
+
+	if ack == nil {
+		sess.pendingPublish[pi] = &mqttPending{
+			jsDur:        jsDur,
+			sseq:         sseq,
+			jsAckSubject: jsAckSubject,
+		}
+	} else {
+		ack.jsAckSubject = jsAckSubject
+		ack.sseq = sseq
+		ack.jsDur = jsDur
 	}
+
 	return pi, dup
 }
 
-// Sends a request to create a JS Durable Consumer based on the given consumer's config.
-// This will wait in place for the reply from the server handling the requests.
+// Stops a PI from being tracked as a PUBLISH. It can still be in use for a
+// pending PUBREL.
 //
 // Lock held on entry
-func (sess *mqttSession) createConsumer(consConfig *ConsumerConfig) error {
-	cfg := &CreateConsumerRequest{
-		Stream: mqttStreamName,
-		Config: *consConfig,
+func (sess *mqttSession) untrackPublish(pi uint16) (jsAckSubject string) {
+	ack, ok := sess.pendingPublish[pi]
+	if !ok {
+		return _EMPTY_
 	}
-	_, err := sess.jsa.createConsumer(cfg)
-	return err
+
+	delete(sess.pendingPublish, pi)
+	if len(sess.pendingPublish) == 0 {
+		sess.last_pi = 0
+	}
+
+	if len(sess.cpending) != 0 && ack.jsDur != _EMPTY_ {
+		if sseqToPi := sess.cpending[ack.jsDur]; sseqToPi != nil {
+			delete(sseqToPi, ack.sseq)
+		}
+	}
+
+	return ack.jsAckSubject
+}
+
+// trackPubRel is invoked in 2 cases: (a) when we receive a PUBREC and we need
+// to change from tracking the PI as a PUBLISH to a PUBREL; and (b) when we
+// attempt to deliver the PUBREL to record the JS ack subject for it.
+//
+// Lock held on entry
+func (sess *mqttSession) trackAsPubRel(pi uint16, jsAckSubject string) {
+	if sess.pubRelConsumer == nil {
+		// The cosumer MUST be set up already.
+		return
+	}
+	jsDur := sess.pubRelConsumer.Durable
+
+	if sess.pendingPubRel == nil {
+		sess.pendingPubRel = make(map[uint16]*mqttPending)
+	}
+
+	if jsAckSubject == _EMPTY_ {
+		sess.pendingPubRel[pi] = &mqttPending{
+			jsDur: jsDur,
+		}
+		return
+	}
+
+	sseq, _, _ := ackReplyInfo(jsAckSubject)
+
+	var sseqToPi map[uint64]uint16
+	if sess.cpending == nil {
+		sess.cpending = make(map[string]map[uint64]uint16)
+	} else if sseqToPi = sess.cpending[jsDur]; sseqToPi == nil {
```

Comment:
Above would clean up quite a lot of things like this.

Reply:
Yes, I even started the change and then decided against it. Maybe a future refactoring, with benchmarks?

Reply:
Yep, let's keep it for a future change then.